### ISSUE TITLE
Send Back to IVR From Listener

### ISSUE THREAD

**Issue by OsborneSupremacy** [bruno222/twilio-flex-plugin-forward-to-ivr#2]
Any suggestions for transferring a call back to IVR from a listener, specifically `afterRejectTask` ?

Here's the listener I'm adding in `init`:

```typescript
flex.Actions.addListener("afterRejectTask", async (payload, abortFunction) => {
    let url: string = payload.task.attributes.transferToIvrUrl;
    let menu: string = 'hangup';
    await request(url, { CallSid: payload.sid, menu });
});
```

Getting this error:

![image](https://user-images.githubusercontent.com/22501712/171544197-2953ad7b-ab1c-40c2-a5dd-b2530966829e.png)

**Comment by bruno222** [comment]
Hi @OsborneSupremacy !

What is the use-case? If one agent rejects the call, the call should go to some other agent available on the queue, no? Curious to understand why you want to send the customer back to the IVR.

Nevertheless, it is possible to archive what you want with this snippet code below:


```
flex.Actions.replaceAction('RejectTask', async (payload, original) => {
      console.log('@@@payload2', payload);

      const url: string = payload.task.attributes.transferToIvrUrl;
      const CallSid: string = payload.task.attributes.call_sid;
      const reservationSid = payload.sid;
      const transferToIVRMenu = 'sales';

      if (!url || !CallSid) {
        await original(payload);
        return;
      }

      await manager.workerClient.reservations.get(reservationSid)?.accept();
      await manager.workerClient.reservations.get(reservationSid)?.wrap();
      await manager.workerClient.reservations.get(reservationSid)?.complete();
      await request(url, { CallSid, transferToIVRMenu });
    });
  }
```

**Comment by OsborneSupremacy** [comment]
@bruno222, that works perfectly. Thank you.

Here's my use case. Bear in mind that I'm new to Flex and my contact center is not yet in production.

In my testing of the contact center, rejecting a call doesn't seem to change the call's status. The call remains in the queue, and there's no change from the caller's perspective.

![DeclineCall](https://user-images.githubusercontent.com/22501712/171857259-37c3c4b5-ebd9-4795-8ed0-a49613c78d3c.gif)

What I'm thinking is that if an agent doesn't want to take the call (or is unavailable), they won't hit the reject button, but let it ring until another agent picks it up, the caller hangs up, or the call is sent to VM automatically after _n_ seconds (that last item is what I plan to tackle next).

If that's not a good pattern, and you could direct me to some resources for configuring Flex / TaskRouter in a better way, I would be grateful!

**Comment by bruno222** [comment]
Cool, glad it worked! But indeed I think you don't need it.

4 topics:

**Topic 1:** The Reject button in your animated gif worked perfectly: You only have one agent available (you!) that is why the task is coming back to you even when you press Reject. Create another Agent (and open Chrome Incognito to have 2 agents in parallel), and you will see the Reject button working properly (having the task jumping from one to another agent)

**Topic 2:** I think you don't need to transfer the call back to IVR. A better way (IMO) would be:

1. Mike the customer calls in
2. Goes to Agent1
3. If Agent1 does not accept the call in 10 seconds, then
         a. the Agent2 from the same queue will receive the Task
         b. and Agent1 is put offline to make sure he does not receive a call anymore... Probably this Agent went away from the computer and forgot to put himself offline
4. If Agent2 does not accept the call in 10 seconds, Agent3 will receive the task... And the loop repeats.

If you like this approach, all you have to do is:

1. In the Console menu, go to TaskRouter > Workflow > select your workflow. At the top, there is an item called "Reservation Timeout" - Set it to 10 seconds.

<img width="751" alt="image" src="https://user-images.githubusercontent.com/1012787/171866889-daa5234b-6fde-41c4-a868-7754c8a4115b.png">

2. Go to menu Settings and make sure your "Timeout Activity" is Offline. This will put the agent offline when he/she is not in front of the computer, once a task times out
 
**Topic 3:**

You said you want to have logic like "after X seconds if no agent is online... I would like to do something, like forwarding the call to VM"

For that, [this might be a good resource](https://docs.google.com/document/d/1hLmYvoqpMapLqay013kVgDzd9qZFQcJm2BRkquHOyFg/edit) for you to learn how to trigger a Function after X seconds of a task is waiting in a queue.

You will have the change the above resource to what you want, so I would say the above is just 50% of the job (to trigger the function after X seconds).

To have a VM working, [this repo](https://github.com/twilio-labs/plugin-queued-callbacks-and-voicemail) does the job "press 1 to record a message". With that, you might be able to copy the logic and, with luck, you will be able to put both solutions together

**Topic 4:** How many agents will you have?

Cheers! Have a great weekend

**Comment by OsborneSupremacy** [comment]
Thanks Bruno, that'll all helpful, and I will explore / consider your recommendations.

I haven't tested with two agents online yet. I probably should have done that earlier!

Initially I'm expecting to have two agents available during business hours, but the number will grow over the course of the year.

**Comment by OsborneSupremacy** [comment]
Hey @bruno222, I went with most of your suggestions. For **Topic 3**, the function approach is working well, for sending a call to VM after no agent picks up after _n_ seconds. I currently have the "send to VM" function being invoked for every _timeout_ event (reservation timeout, task queue timeout, workflow timeout).

One edge case I'm wondering how to handle is when no agents are online when the call comes in. In that case, no timeout event is getting triggered. Would there be a different event that's triggered in that scenario? I tried "Reservation Failed," but that didn't do it.

**Edit:**

I decided on a different approach -- creating a serverless function to query the taskrouter workspace to see if there are any available agents. If no agents are available, I don't transfer to Flex, but right to VM. So, I no longer need this question answered. Thanks again for your help!
